contrib/dump_btrfs_bgs.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env drgn
+# Copyright (c) Western Digital Corporation, Inc. and affiliates.
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""  Dump all block group caches for a given btrfs file-system """
+
+import sys
+import drgn
+from enum import Flag
+from drgn import NULL, Object, cast, container_of, execscript, \
+        reinterpret, sizeof
+from drgn.helpers.linux import *
+
+class BtrfsBlockGroupFlags(Flag):
+    BTRFS_BLOCK_GROUP_DATA      = (1 << 0)
+    BTRFS_BLOCK_GROUP_SYSTEM    = (1 << 1)
+    BTRFS_BLOCK_GROUP_METADATA  = (1 << 2)
+    BTRFS_BLOCK_GROUP_RAID0     = (1 << 3)
+    BTRFS_BLOCK_GROUP_RAID1     = (1 << 4)
+    BTRFS_BLOCK_GROUP_DUP       = (1 << 5)
+    BTRFS_BLOCK_GROUP_RAID10    = (1 << 6)
+    BTRFS_BLOCK_GROUP_RAID5     = (1 << 9)
+    BTRFS_BLOCK_GROUP_RAID6     = (1 << 8)
+    BTRFS_BLOCK_GROUP_RAID1C3   = (1 << 9)
+    BTRFS_BLOCK_GROUP_RAID1C4   = (1 << 10)
+
+class BtrfsBlockGroupRuntimeFlags(Flag):
+    BLOCK_GROUP_FLAG_IREF                   = (1 << 0)
+    BLOCK_GROUP_FLAG_REMOVED                = (1 << 1)
+    BLOCK_GROUP_FLAG_TO_COPY                = (1 << 2)
+    BLOCK_GROUP_FLAG_RELOCATING_REPAIR      = ( 1 << 3)
+    BLOCK_GROUP_FLAG_CHUNK_ITEM_INSERTED    = (1 << 4)
+    BLOCK_GROUP_FLAG_ZONE_IS_ACTIVE         = (1 << 5)
+    BLOCK_GROUP_FLAG_ZONED_DATA_RELOC       = ( 1 << 6)
+    BLOCK_GROUP_FLAG_NEEDS_FREE_SPACE       = ( 1 << 7)
+    BLOCK_GROUP_FLAG_SEQUENTIAL_ZONE        = (1 << 8)
+
+if len(sys.argv) > 1:
+    mnt_path = sys.argv[1]
+    mnt_path = mnt_path.rstrip('/')
+else:
+    mnt_path = "/"
+
+mnt = None
+
+for mnt in for_each_mount(prog, dst = mnt_path):
+    pass
+
+if mnt is None:
+    sys.stderr.write(f'Error: mount point {mnt_path} not found')
+    sys.exit(1)
+
+try:
+    fs_info = cast('struct btrfs_fs_info *', mnt.mnt.mnt_sb.s_fs_info)
+except LookupError:
+    print('cannot find \'struct btrfs_fs_info *\', module not loaded?')
+    sys.exit(1)
+
+def dump_bg(bg):
+    print(f'BG at {bg.start.value_()}')
+    print(f'\tflags: {str(BtrfsBlockGroupFlags(bg.flags.value_())).split(".")[1]} ({hex(bg.flags)})')
+    print(f'\tlength: {bg.length.value_()}')
+    print(f'\tused: {bg.used.value_()}')
+    print(f'\tpinned: {bg.pinned.value_()}')
+    print(f'\treserved: {bg.reserved.value_()}')
+    print(f'\truntime_flags: {str(BtrfsBlockGroupRuntimeFlags(bg.runtime_flags.value_())).split(".")[1]} ({hex(bg.runtime_flags)})')
```

Comment:
The `enum.Flag` stuff failed for me on Python 3.11:

```
bash-5.2# python3 -m drgn ./contrib/dump_btrfs_bgs.py  /mnt
BG at 13631488
        flags: BTRFS_BLOCK_GROUP_DATA (0x1)
        length: 8388608
        used: 0
        pinned: 0
        reserved: 0
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/osandov/repos/drgn2/drgn/__main__.py", line 16, in <module>
    _main()
  File "/home/osandov/repos/drgn2/drgn/cli.py", line 225, in _main
    runpy.run_path(script, init_globals={"prog": prog}, run_name="__main__")
  File "<frozen runpy>", line 291, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "./contrib/dump_btrfs_bgs.py", line 73, in <module>
    dump_bg(bg)
  File "./contrib/dump_btrfs_bgs.py", line 66, in dump_bg
    print(f'\truntime_flags: {str(BtrfsBlockGroupRuntimeFlags(bg.runtime_flags.value_())).split(".")[1]} ({hex(bg.runtime_flags)})')
                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

I suspect because of changed behavior of `str(enum.Flag)` in Python 3.11:

```
$ cat test.py
import enum

class Foo(enum.Flag):
    X = 1
    Y = 2

print(str(Foo(0)))
$ python3.10 test.py
Foo.0
$ python3.11 test.py
Foo(0)
```

You could use [`drgn.helpers.common.decode_flags()`](https://drgn.readthedocs.io/en/latest/helpers.html#drgn.helpers.common.format.decode_flags) instead of `enum.Flag` to work around this.

Reply:
Oh cool, didn't know about that one. I'll update :)